### DISCUSSION TITLE
ES6 module update

### DISCUSSION THREAD

**Discussion by ghost** [boytchev/mannequin.js#32] [General]
I made a huge technical update to the mannequin.js codebase:
 
- refactored the code to ES6 module structure, changing file endings to .mjs
- decoupled the lib and the scene setup - no global vars and dependencies any more
- set up a default code style (standard js) and fixed the code problems, added some JSDoc comments
- added a package.json and a simple npm based install procedure for three.js and example artifacts needed
- updated the codebase to the latest version 160 of three.js by fixing the deprecated light code
- updated all examples to use the new, module imports based interface

As mentioned [here](https://github.com/boytchev/mannequin.js/issues/30), the posture editor is now a web component.
As proof it is instantiated 3 times side by side in editor/index.html :)

All changes to this proposed new major version 5.0 can be reviewed [here](https://github.com/lavatroll/mannequin.js).

I did all this as needed for my project and publishing it as required by the - very welcome - free software license. 

But at the same time I consider this an important step forward for this ingeniously crafted and artistically designed piece of software. The package, in this state, can be published to npm with a simple "npm publish“ and consumed right away, as any build tool will recognize the ES6 modules.

Would love to hear comments and ideas for further improvement :)

**Comment by ghost** [comment]
After finally learning about Euler angles I also fixed the annoying rotation changes in unrelated dimensions.
The order is now always XYZ, no side effects in getters and setters any more. 
Is there anything I have overlooked with this change?

**Comment by boytchev** [comment]
Thanks for your work. This is just to let you know I'm aware of your suggestions. But I need more time to reach to resolution, because:
* currently I'm too busy, I hope in April I might have some time
* the proposed changes are in conflict with my current use of the library (it is in an old-fashioned environment), so I'd move the code into a more modern style only if I find a way how to keep its backward usability

**Comment by ghost** [comment]
Thanks for making your process clear and for considering my changes.

I fully understand that for university teaching, the old fashioned JS style with global vars, functions and classes is easier to read and write. At the same time, modules and imports is how JavaScript is used today. 

I propose a compatibility wrapper around the new mannequin.mjs, making the old use via the global namespace again possible, e.g. like this:

```
import * as mannequin from '../mannequin.mjs'
import * as label from './fonts/label.mjs'
Object.assign(window, mannequin, label)
```
